README
===== Zim - A desktop wiki and outliner =====

=== Install ===

see [[./share/zim/doc/zim/install]]

=== Copyright ===

Copyright (c) 2005, 2008 Jaap G Karssenberg. All rights reserved.
This program is free software; you can redistribute it and/or
modify it under the same terms as Perl.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== Included Files ==

The files in [[./contrib/]] are included from third party contributors.
Also the following files were included from other sources:

From the default Gnome icon theme:
* share/pixmaps/zim/stock_todo.png
* share/pixmaps/zim/mail-attachment.png

From Gtk+ 2.8
* share/pixmaps/zim/gtk-connect.png (was: stock_connect_24.png)

Other:
* share/pixmaps/zim/stock_calendar-view-month.png
	Copyright 2007 by Jakub Steiner, released under GPL
	modifications copyright 2009 by Gabriel Hurley

=== Website ===

http://zim-wiki.org

=== Bugs ===

Please report bugs to https://bugs.launchpad.net/zim/ or mail pardus@cpan.org



Changes
===== Changes for zim =====
Jaap Karssenberg pardus@cpan.org

=== 0.29 - Sun Jan 10 2010 ===
* Suport tags in the TODOList
* Improve date parsing for TODOList
* Add option for upper-case formatting for page titles in new pages
* Punctuation triggers end-of-word for auto-formatting links
* Updated website in About dialog
* Added translation for Hungarian, Japanese, Brazilian Portuguese & Ukrainian

=== 0.28 - Tue Feb 17 2009 ===
* Reversed behavior of html export to use <br> elements - lp #311292
* Fixed install script for Module::Build version 0.3x - lp #319964
* Added support for RTL languages in html export - lp #303108
* Added DiagramEditor plugin to insert Graphviz diagrams
* Added SortSelection plugin - lp #318904
* Applied patch to improve perl Pod format backend by Christoph Rudorff
* Applied patch to set toolbar style by Christoph Rudorff
* Added calendar icon with tango colors, submitted by Gabriel Hurley
* Changed EquationEditor latex template to allow line breaks - lp #301658
* Fixed two syntax bugs - lp #297932 & lp #312894
* Fixed bug where checkboxlist with an empty line causes a freeze - lp #300184
* Fixed bug where page headers show in "send to.." email content - lp #316153
* Replaced horizontal scrollbar for page index by ellipsize - lp #297682

=== 0.27 - Wed Nov 12 2008 ===
* Added translation for traditional Chinese
* Fixed Subversion support same level as Bazaar - lp #263043
* Added support for checkbox lists - lp #283901
* Added feature to automount notebooks - e.g. to use fuse encfs
* Fixed bug that the Bazaar test did not skip properly if bzr not installed
* Fixed bugs for using gjots notebooks
* Fixed bug open notebooks were not italized in notebook dialog - lp #272518
* Fixed bug with toggling format buttons - lp #273577
* Added tooltips to pathbar buttons - lp #273593
* Changed window title to notebook name instead of page name - lp #272430

=== 0.26 - Sun Aug 24 2008 ===
* Added translations for Finnish, Turkish and Simplified Chinese
* Added integrated version control support for zim notebooks based on Bazaar.
  Added "Save Version" (<Ctrl><Shift>S) and "Versions" dialog to File menu.
* Zim text files now have headers with content type and time stamps
* Changed wiki syntax for Verbatim blocks, now allows non-verbatim paragaphs
  and bullet lists to be indented - lp #208986
* Added mimetype for zim text files text/x-zim-wiki
* Changed mimetype for .zim files to application/x-zim-notebook
* Added special notebook for man pages with "--man" commandline option
* Added support for reading gjots2(1) files; added application/x-gjots mimetype
* Made auto incrementing a numbered list optional - lp #235020
* Added translation support for plural forms - lp #226976
* Fixed bugs for perl 5.10 - lp #235030, #242490
* Fixed bug in Calendar plugin when namespace is read-only
* Fixed bug in ToggleReadOnly plugin when first page is read-only
* Fixed bug with TrayIcon for single notebook mode - lp #252712
* Fixed bug that prevented typing under win32 - lp #252005

=== 0.25 - Tue May 20 2008 ===
* Added translations for Danish, Russian and Galician
* Added --gtk-path to work around win32 environment behavior
* Fixed a bug when opening files without asociated application
* Fixed a bug for older versions of Exporter
* Fixed bug that prevented using Gtk2::StatusIcon this means we can
  drop dependency on Gtk2::TrayIcon for gtk+ >= 2.10
* Fixed bug with utf8 notebook name not showing correctly in TrayIcon menu
* Fixed TrayIcon to hide window when it already has focus
* Fixed bug with file selection dialog for certain gtk versions
* Fixed bug where a few strings were skipped when indexing translatable text

=== 0.24 - Mon Apr 21 2008 ===
* Added translations for German, French, Italian and Hebrew
* Improved support for attaching files and images
* Added daemon component to make seperate instances communicate
* Added a TODO list plugin that searches for TODO tags in the notebook
* Added a ToggleReadOnly plugin to make a notebook temporary read-only
* Moved Zim::File to Zim::FS::File, also added Zim::FS::Dir objects
  made whole file system access much more object oriented
* Config files now use proper INI-style layout
* Notebook config file is now called "notebook.zim" and has it's own mimetype
* The history, cache and state are now saved under ".zim/" in the notebook
* Changed directory structure of the manual
* Moved Calendar functionality back into plugin
* Calendar now namespaces date pages per month for better scalability.
  On first use zim will prompt to move exiting pages.
* Date namespace now has it's own template for new pages
* Added a keybinding ( ">" ) to quote selections
* Updated Build script, we now need Module::Build >= 0.28
* Website changed, zim now lives at: http://zim-wiki.org

=== 0.23 - Wed Nov 21 2007 ===
* Added translation support - translators welcome !
* Added translations for Dutch, Spanish, Czech, Swedish and Portuguese
* Renamed "repository" to "notebook" for better conceptual mapping
* Fixed bug when writing to symlinked files
* Fixed bug with "Edit Link" action on unlinked text
* Fixed bug in conversion of link text in exported HTML
* Added key bindings to side pane: '\' and '*' for collapse and expand all
* Now requires File::BaseDir 0.03 and File::DesktopEntry 0.03 or better

=== 0.22 - Thu Oct 11 2007 ===
Bug fix release :(
* Fixed bug that prevented the creation of pages in new namespaces by
  labelling them as read-only
* Added current format style to status bar
* Fixed bug with "Send To..." menu item

=== 0.21 - Sun Sep 30 2007 ===
* Added an EquationEditor plugin using latex and dvipng
* Added a InsertScreenshot plugin using scrot(1) - idea from Armin Taschwer
* Added a "Open in external editor" menu item
* Made read_only property switch per page
* Fixed bug with the file name for the home page
* Fixed bug with utf8 in template for new pages

=== 0.20 - Sun Aug 19 2007 ===
* Added config file to control display styles of text formatting

Build.PL
#!/usr/bin/perl

use Module::Build;

# Define our custom build actions
my $class = Module::Build->subclass(
	class => 'My::Builder',
	code => << 'EOC'

use strict;
use File::Copy ();
use File::Path;

$ENV{PERL5LIB} ||= ''; # get rid of some warnings...

__PACKAGE__->add_property(data_files => {});

sub check_prereq {
	my $self = shift;

	eval 'use Gtk2';
	my $gtk_ok = ! $@;

	if ($gtk_ok and Gtk2->CHECK_VERSION(2, 10, 0)) {
		my $recommends = $self->recommends;
		delete $recommends->{'Gtk2::TrayIcon'};
	}

	my $prereq_ok = $self->SUPER::check_prereq(@_);
	
	if (! $gtk_ok) {
		$self->log_warn(" * ERROR: Could not load perl module Gtk2\n");
		$self->log_warn(<<EOF) if $prereq_ok;

ERRORS/WARNINGS FOUND IN PREREQUISITES.  You may wish to install the versions
of the modules indicated above before proceeding with this installation

EOF
	}
	elsif (! Gtk2->CHECK_VERSION(2, 4, 0)) {
		my $version = join '.', Gtk2->GET_VERSION_INFO;
		$self->log_warn(<<"EOF");

 * ERROR: gtk+ version $version is installed, but we need >= 2.4.0
   If you think you have the right version of gtk+ installed, try
   recompiling the Gtk2 perl package.

EOF
	}

	return $prereq_ok && $gtk_ok;
}

sub install_types {
	my $self = shift;
	my @types = $self->SUPER::install_types;
	push @types, keys %{$$self{properties}{data_files}};
	return sort @types;
}

sub install_destination {
	my ($self, $type) = @_;
	return $self->SUPER::install_destination($type)
		unless exists $$self{properties}{data_files}{$type};

	return $self->install_path($type) if $self->install_path($type);

	my $relpath = $$self{properties}{data_files}{$type};

	return File::Spec->catdir($self->install_base, $relpath)
		if $self->install_base;

	return File::Spec->catdir($self->prefix, $relpath)
		if $self->prefix;
	
	# Fall back to heuristic determination of prefix based on
	# the 'bin' directory.
	my $bindir = $self->install_sets($self->installdirs)->{'bin'};
	my @prefix = File::Spec->splitdir($bindir);
	pop @prefix; # loose 'bin'
	return File::Spec->catdir(@prefix, $relpath);
}

sub process_share_files {
	my $self = shift;
	my $blib = $self->blib;

	# Copy but exclude dev documentation
	for (@{$self->rscan_dir('share')}) {
		next if /(^|[\/\\])\.|Development/ or -d $_;
		$self->copy_if_modified(
			from => $_,
			to   => File::Spec->catfile($blib, $_) );
	}
	
	# Add changelog
	my $dir = File::Spec->catdir($blib, qw/share zim doc/);
	$self->copy_if_modified(
		from => 'Changes',
		to   => File::Spec->catfile($dir, 'changelog.txt'));

	# Set correct config for manual
	my $conf = File::Spec->catfile($dir, 'notebook.zim');
	open CONF, '<', $conf or die $!;
	my $text = join '', <CONF>;
	close CONF;
	
	if ($text =~ s/^#(?!\s)//mg) { # un-escape some lines
		my $mode = (stat $conf)[2];
		chmod($mode | oct(222), $conf) or die $!;
		print "Making the manual a read-only notebook\n";
		open CONF, '>', $conf or die $!;
		print CONF $text;
		close CONF;
		chmod($mode, $conf);
	}
}

sub ACTION_install {
	my $self = shift;
	$self->SUPER::ACTION_install(@_);
	$self->depends_on('postinstall');
}


sub ACTION_test {
	my $self = shift;

	# Multiple tests depend on data from the MANIFEST, so it needs to 
	# be correct. However only do this in dev tree when we can actually
	# rebuild the manifest.
	if (-e 'MANIFEST.SKIP') {
		my $error = 0;
		local $SIG{__WARN__} = sub {$error++; warn @_};
		$self->depends_on('distcheck');
		die "Please update the MANIFEST before testing\n" if $error;
	}

	$self->depends_on('test_data');
	$self->SUPER::ACTION_test(@_);
}

sub ACTION_test_data {
	my $self = shift;

	for (qw#t/html t/notebook t/config t/share t/cache#) {
		next unless -e $_;
		$self->delete_filetree($_);
	}

	print "Extracting test notebook\n";

	open IN, 't/notebook.txt' or die "t/notebook.txt: $!";
	binmode IN, ':utf8' if not $] < 5.008;

	my ($file, $fh);
	while (<IN>) {
		if (/^%%\s+(\S+)\s+%%$/) {
			close $fh if $fh;
			$file = $1;
			mkpath $1 if $file =~ m#(.*/)#;
			open $fh, ">$file" or die "$file: $!\n";
			binmode $fh, ':utf8' if not $] < 5.008;
		}
		elsif ($fh) {
			print $fh $_;
		}
		else { die "oops" }
	}
	close $fh if $fh;
}

#sub ACTION_distmeta {
	#my $self = shift;
	#$self->depends_on('messages');
	#$self->run_perl_script('Lingua.PL') or return;
	#$self->run_perl_script('Config.PL') or return;
	#$self->SUPER::ACTION_distmeta(@_);
#}

sub ACTION_disttest {
	my $self = shift;
	$self->SUPER::ACTION_disttest(@_);

	# Check how clean we build
	my $start_dir = $self->cwd;
	my $dist_dir = $self->dist_dir;
	chdir $dist_dir or die "Cannot chdir to $dist_dir: $!";
	$self->run_perl_script('Build', [], ['realclean']);
	chdir $start_dir;

	my %manifest;
	open MANIFEST, 'MANIFEST' or die "Cannot open MANIFEST: $!";
	while (<MANIFEST>) {
		chomp;
		s/\s.*//;
		$manifest{$_}++;
	}
	close MANIFEST;
	my @dirt;
	my @files = grep {-f $_} @{$self->rscan_dir($dist_dir)};
	for (@files) {
		s/^$dist_dir[\/\\]//;
		push @dirt, $_ unless $manifest{$_};
	}
	die "ERROR: The following files where not cleaned up properly:\n",
		map "\t$_\n", @dirt
		if @dirt;
	print "Cleanup OK\n";
}

sub ACTION_realclean {
	my $self = shift;
	$self->SUPER::ACTION_realclean(@_);
	my @files = grep /(\~|\.bak)$/, @{$self->rscan_dir('.')};
	return unless @files;
	warn "Also cleaning up:\n", map "\t$_\n", @files;
	unlink $_ or die $! for @files;
}

=head1 ACTIONS

=over 4

=item C<postinstall>

Runs commands to update the XDG databases.

=cut

sub ACTION_postinstall {
	my $self = shift;
	my $sharedir = $self->install_destination('share');
	my $mimedir = File::Spec->catdir($sharedir, 'mime');
	my $icon = File::Spec->catfile($sharedir, 'pixmaps', 'zim.png');
	for (
		['update-desktop-database'],
		['update-mime-database', $mimedir],
	#	[qw{xdg-icon-resource install --context apps --size 64},
	#		$icon],
		[qw{xdg-icon-resource install --context mimetypes --size 64},
			$icon, q{application-x-zim-notebook}],
	) {
		print "Trying: @$_\n";
		print system(@$_) == 0 ? "Ok\n\n" : "Failed\n\n" ;
	}

}

=item C<messages>

Generates a new F<zim.pot> file. Only works from development tree,
not from dist.

=cut

sub ACTION_messages {
	my $self = shift;
	die "This is not a development tree, can not build messages.\n"
		unless -d './dev';
	$self->do_system($self->perl, './dev/xgettext.pl');
}

=item C<static>

Builds a self-contained excutable for zim. Needs the L<PAR> and
L<ExtUtils::PkgConfig> modules.

=back

=cut

sub ACTION_static {
	my $self = shift;
	warn "## WARNING: Static building is experimental - feedback is welcome\n";
	warn "## You need to have zim installed before running this command\n";
	warn "## You can pass extra options to 'pp' using PP_OPTS\n";
	eval 'use ExtUtils::PkgConfig';
	die $@ if $@;
	sleep 5;
	$self->depends_on('build');
	my @libs = map {s/^-l//i; ('-l', $_)} grep {length $_} split /\s+/,
		ExtUtils::PkgConfig->libs('gtk+-2.0');
	my @add = ('-a', 'share');
	open MANIFEST, '<MANIFEST' or die "Could not read MANIFEST";
	my @modules;
	for (<MANIFEST>) { # grep modules from the manifest
		/^lib\/(\S+?)\.pm/ or next;
		my $mod = $1;
		$mod =~ s/\//::/g;
		push @modules, '-M', $mod;
	}
	close MANIFEST;
	my $script = File::Spec->catfile(qw/script zim/);
	my $blib = File::Spec->catdir($ENV{PWD}, $self->blib);
	my $libdir = File::Spec->catdir($blib, 'lib');
	print ">> cd $blib\n";
	chdir $blib or die "Could not open dir $blib\n";
	$ENV{PWD} = $blib;
	print ">> pp -I $libdir -o zim @libs @add @modules $script\n";
	system qw{pp -o ../zim}, @libs, @add, @modules, $script;
	warn "\n## Your executable is ready\n";
}

#=item C<testflow>
#
#Run test suite and record which method calls which other methods.
#
#=cut

sub ACTION_testflow {
	# TODO: Move "tmon.out" after each test, now it gets overwritten
	# TODO: wrap script around "dprofpp -T" to make summaries or graphs
	my $self = shift;

#	$self->add_to_cleanup('flow_db');
	$self->depends_on('code');

	local $Test::Harness::switches    =
	local $Test::Harness::switches    =
	local $ENV{HARNESS_PERL_SWITCHES} = '-d:DProf';

	$self->depends_on('test');
#	$self->do_system('./dev/flow.pl');
}

EOC
);

# Define the build parameters
my $build = $class->new(
	dist_name => 'Zim',
	dist_version_from => 'lib/Zim.pm',
	dist_abstract => 'A desktop wiki',
	dist_author => 'Jaap Karssenberg <pardus@cpan.org>',
	license => 'perl',
	script_files => ['bin/zim'],
	data_files => { share => 'share' },
	requires => {
		'perl'           => '5.8.0',
		'Gtk2'           => '1.040',
		# Older doesn't support Gtk2::UIManager I think
		'POSIX'          => 0,
		'IO::File'       => 0,
		'File::Spec'     => 0,
		'File::Copy'     => 0,
		'File::BaseDir'  => 0.03,
		'File::MimeInfo' => '0.12',
		'File::DesktopEntry' => 0.03,
		'Encode'         => 0,
	},
	recommends => {
		'Gtk2::TrayIcon' => 0,
			# This recommendation is removed on the
			# fly when gtk+ >= 2.10 is detected because
			# in that case we can use Gtk2::StatusIcon
		'Gtk2::Spell'    => 0,
	},
	build_requires => {
		'Module::Build'  => '0.28',
		'Test::More'     => 0,
		'File::Path'     => 0,
	},
	add_to_cleanup => [
		'./cover_db/',	# from Build testcover
		'./tmon.out',	# from dprof
		'./tmp',
		# Used by tests:
		't/notebook/',
		't/html/',
		't/config/',
		't/share/',
		't/cache/',
	],
	create_makefile_pl => 'passthrough',
	dynamic_config => 1,
);

# Add share dir as something to be build
$build->add_build_element('share');

# And finally create the script "./Build"
$build->create_build_script;

